src/preprocessing/passes/ackermann.cpp

```diff
@@ -187,9 +188,110 @@ void collectFunctionsAndLemmas(FunctionToArgsMap& fun_to_args,
 
 /* -------------------------------------------------------------------------- */
 
+/* Given a minimum capacity for an uninterpreted sort, return the size of the
+ * new BV type */
+size_t getBVSkolemSize(size_t capacity)
+{
+  return static_cast<size_t>(log2(capacity)) + 1;
+}
+
+/* Given the lowest capacity requirements for each uninterpreted sort, assign
+ * a sufficient bit-vector size. Get the converting map */
+void collectUSortsToBV(const vector<TNode>& terms,
+                       USortToBVSizeMap& usortCardinality,
+                       SubstitutionMap& sortsToSkolem)
+{
+  NodeManager* nm = NodeManager::currentNM();
+
+  for (TNode term : terms)
+  {
+    TypeNode type = term.getType();
+    size_t size = getBVSkolemSize(usortCardinality[type]);
+    Node skolem = nm->mkSkolem(
+        "BVSKOLEM$$",
+        nm->mkBitVectorType(size),
+        "a variable created by the ackermannization "
+        "preprocessing pass, representing a term with uninterpreted sort.");
+    sortsToSkolem.addSubstitution(term, skolem);
+  }
+}
+
+/* This function returns the list of terms with uninterpreted sort in the
+ * formula represented by assertions. */
+std::vector<TNode> getVarsWithUSorts(AssertionPipeline* assertions)
+{
+  TNodeSet seen;
+  std::vector<TNode> res;
+
+  for (const Node& assertion : assertions->ref())
+  {
+    std::unordered_set<TNode, TNodeHashFunction> vars;
+    expr::getVariables(assertion, vars);
+
+    for (const TNode& var : vars)
+    {
+      if (var.getType().isSort() && seen.find(var) == seen.end())
+      {
+        seen.insert(var);
+        res.push_back(var);
+      }
+    }
+  }
+
+  return res;
+}
+
+/* This is the top level of converting uninterpreted sorts to bit-vectors.
+ * We count the number of different variables for each uninterpreted sort.
+ * Then for each sort, we will assign a new bit-vector type with a sufficient
+ * size. The size is calculated to have enough capacity, that can accommodate
+ * the variables occured in the original formula. At the end, all variables of
+ * uninterpreted sorts will be converted into Skolem variables of BV */
+void usortsToBitVectors(const LogicInfo& d_logic,
+                        AssertionPipeline* assertions,
+                        USortToBVSizeMap& usortCardinality,
+                        SubstitutionMap& sortsToSkolem)
+{
+  std::vector<TNode> toProcess = getVarsWithUSorts(assertions);
+
+  if (toProcess.size() > 0)
+  {
+    /* the current version only supports BV for removing uninterpreted sorts */
+    AlwaysAssert(d_logic.isTheoryEnabled(theory::THEORY_BV))
+        << "Cannot use Ackermannization on formula with uninterpreted "
+           "sorts without BV logic";
+
+    for (TNode term : toProcess)
+    {
+      /* ackermannizatino assumes no quantifiers */
+      AlwaysAssert(term.getKind() != kind::BOUND_VARIABLE)
```

Comment:
I think this check should be done at the beginning of the entire Ackermannization preprocessing pass.

Reply:
I'd suggest that we just check in `SmtEngine` whether the logic is quantifier-free or not. What do you think about that?